README.md
<p align="center">
  <img src="/.github/images/banner.png?sanitize=true" height="100" />
</p>

<p align="center">
  <img src="https://img.shields.io/github/package-json/v/theca11/whack-a-mole/master?label=%20&color=green" />
  <a href="https://marketplace.elgato.com/product/whack-a-mole-c322a2db-bcdb-4531-ac6c-28ec05a3372d"><img src="https://img.shields.io/badge/Elgato_Marketplace-gray?logo=elgato&logoColor=white&labelColor=1231ac&color=gray" alt="Elgato Marketplace" /></a>
  <a href="https://discord.gg/yFce5NEDak"><img src="https://img.shields.io/badge/Discord-gray?logo=discord&logoColor=white&labelColor=6A7EC2&color=gray" alt="Discord Support" /></a>
  <a href="https://ko-fi.com/the_ca11"><img src="https://img.shields.io/badge/Ko--fi-gray?logo=kofi&logoColor=white&labelColor=red&color=gray" alt="Ko-fi donate" /></a>
</p>

<hr/>

Whack-A-Mole plugin for Elgato Stream Deck lets you play the popular game.

## Game Instructions
The goal is to whack as many moles as possible before the time runs out.

There are 3 types of moles: regular moles add 1 to the score, shy moles add 2 since they're harder to spot, and miner moles substract 1 because they have protective helmets! In endless mode the same applies when adding seconds to the timer.

How to:
- **Whack**: press the key with the mole to whack it
- **Start**: press to start game, long press to cancel an ongoing game
- **Set Difficulty/timer**: press to change game length (5s/10s/20s/30s/endless), long press to change game difficulty (easy/normal/hard/custom)
- **Show Score**: press to toggle between last and top score for the current difficulty/length

### Custom Difficulty Level
A custom difficulty level can be configured and selected. To do so, click on any of the plugin's keys and open the external window using the provided link. You'll find there some input fields. The values to input refer to the min and max time in milliseconds that a mole can peep for. Lower values mean faster moving moles. For reference, the normal difficulty level settings are 650-950

### Top Scores
Top scores for every difficulty/mode are saved so you can try to beat them. You can view your top scores either by pressing the score key, which toggles between last score and top score, or by opening the external window linked in the keys configuration

## Installation

### Installation from Elgato Marketplace (recommended)

Go to the [Elgato Marketplace plugin page](https://marketplace.elgato.com/product/whack-a-mole-c322a2db-bcdb-4531-ac6c-28ec05a3372d) and click the `Get` button. Follow the steps to download the plugin.

### Installation from Github release

Go to the [releases page](https://github.com/theca11/whack-a-mole/releases) and choose your desired version. In the `Assets` section at the end of the release post you will find a file `Whack-A-Mole-Plugin-v#.#.#.streamDeckPlugin`. Download it and double-click it to make Stream Deck prompt for installation confirmation.

Alternatively, you can download the zipped build folder also found in the release `Assets` (`Whack-A-Mole-Plugin-v#.#.#.zip`), extract it, and copy/paste the folder `dev.theca11.whack-a-mole.streamDeckPlugin` to the corresponding Stream Deck Plugins directory (for Windows located at `%APPDATA%\Elgato\StreamDeck\Plugins`).

> Official Elgato docs on installing/uninstalling plugins [here](https://help.elgato.com/hc/en-us/articles/11434818801293-Elgato-Stream-Deck-How-to-Install-and-Uninstall-Stream-Deck-Plugins-).


## FAQ

**> How can I report an issue, make a feature request or get general help?**

Use [Github](https://github.com/theca11/whack-a-mole/issues) for unexpected issues and feature requests, or join the [Discord server](https://discord.gg/yFce5NEDak) to ask general questions and get help.


**> I'm using your plugin and I love it! How can I show appreciation?**

You can stop by Discord to share your love. If you want to support the project financially, you can [buy me a coffee](https://ko-fi.com/the_ca11).


CHANGELOG.md
# CHANGELOG

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2023-11-22

### Added

-   Added action to switch to the game profile. If you use this action, you can then exit the game by holding the Start/Exit button, which will return you to the previous profile

### Changed

-   Start/Timer/Score/Tile actions no longer visible in actions list

## [1.0.1] - 2023-09-30

### Fixed

-   Added actions to action list so the plugin can be uninstalled from there.
-   Properly set/reset actions state/title when they appear.

## [1.0.0] - 2023-09-28

First release of Whack-A-Mole Stream Deck plugin, which lets you play the popular game.

The goal is to whack as many moles as possible before the time runs out.

There are 3 types of moles: regular moles add 1 to the score, shy moles add 2 since they're harder to spot, and miner moles substract 1 because they have protective helmets! In endless mode the same applies when adding seconds to the timer.

**How to**

-   **Whack**: press the key with the mole to whack it
-   **Start**: press to start game, long press to cancel an ongoing game
-   **Set Difficulty/timer**: press to change game length (5s/10s/20s/30s/endless), long press to change game difficulty (easy/normal/hard/custom)
-   **Show Score**: press to toggle between last and top score for the current difficulty/length

**Custom Difficulty Level**

A custom difficulty level can be configured and selected. To do so, click on any of the plugin's keys and open the external window using the provided link. You'll find there some input fields. The values to input refer to the min and max time in milliseconds that a mole can peep for. Lower values mean faster moving moles. For reference, the normal difficulty level settings are 650-950

**Top Scores**

Top scores for every difficulty/mode are saved so you can try to beat them. You can view your top scores either by pressing the score key, which toggles between last score and top score, or by opening the external window linked in the keys configuration


webpack.config.js
const TerserPlugin = require('terser-webpack-plugin');
const CopyPlugin = require('copy-webpack-plugin');
const LicenseWebpackPlugin = require('license-webpack-plugin').LicenseWebpackPlugin;

module.exports = {
	context: __dirname,
	entry: {
		app: './src/plugin/app.ts',
	},
	resolve: {
		extensions: ['.ts', '.js'],
	},
	module: {
		rules: [
			{
				test: /\.ts$/,
				use: 'ts-loader',
				exclude: /node_modules/,
			},
		],
	},
	target: 'web',
	output: {
		filename: 'plugin/[name].js',
		path: __dirname + '/build/dev.theca11.whack-a-mole.sdPlugin',
		clean: true,
	},
	optimization: {
		concatenateModules: false,
		minimizer: [
			new TerserPlugin({
				test: /app\.js(\?.*)?$/i,
				terserOptions: {
					ecma: 2020,
					mangle: false,
					keep_classnames: true,
					keep_fnames: true,
					module: false,
				},
			}),
		],
	},
	plugins: [
		new CopyPlugin({
			patterns: [
				{
					from: 'src',
					globOptions: {
						ignore: ['**/*.ts', '**/.*'],
					},
				},
				'CHANGELOG.md',
				'LICENSE',
				'README.md',
			],
		}),
		new LicenseWebpackPlugin({
			outputFilename: '3rdparty-licenses.txt',
			renderLicenses: (modules) =>
				modules
					.map(
						(module) =>
							`${module.packageJson.name} (v${module.packageJson.version}) - ${module.licenseId}\n${module.licenseText}`,
					)
					.join('\n\n-------------------------\n\n'),
		}),
	],
	watchOptions: {
		ignored: /node_modules/,
	},
};


src/plugin/types.d.ts
// --- StreamDeck lib types for events ---
type Common = {
	action: string,
	event: string,
	context: string,
	device: string,
}
type Payload<T> = {
	payload: {
		coordinates: { column: number, row: number },
		isInMultiAction: boolean,
		state?: number,
		settings: T
	}
}
type TriggerPayload<T> = Payload<T> & { payload: { userDesiredState?: number } }
type BaseEventData<T> = Common & Payload<T>
type TriggerEventData<T> = Common & TriggerPayload<T>

export type DidReceiveSettingsData<T> = BaseEventData<T>;
export type DidReceiveGlobalSettingsData<T> = {
	event: string,
	payload: { settings: T }
}
export type KeyDownData<T> = TriggerEventData<T>;
export type KeyUpData<T> = TriggerEventData<T>;
export type WillAppearData<T> = BaseEventData<T>;
export type WillDisappearData<T> = BaseEventData<T>;
export type SendToPluginData<T> = {
	action: string,
	event: string,
	context: string,
	payload: T
}
export type SendToPIData<T> = {
	action: string,
	event: string,
	context: string,
	payload: T
}
// ---

src/plugin/SDAction.ts
import { KeyDownData, KeyUpData } from './types';

export class SDAction extends Action {
	_pressCache = new Map<string, NodeJS.Timeout>(); // <context, timeoutRef>

	constructor(UUID: string) {
		super(UUID);

		this.onKeyDown((evtData: KeyDownData<unknown>) => {
			const { context } = evtData;

			if (this._pressCache.has(context)) return;

			const timeout = setTimeout(() => {
				this._pressCache.delete(context);
				this.emit(`${this.UUID}.longPress`, evtData);
			}, 300);
			this._pressCache.set(context, timeout);
		});

		this.onKeyUp((evtData: KeyUpData<unknown>) => {
			const { context } = evtData;
			if (!this._pressCache.has(context)) {
				return;
			}
			else {
				clearTimeout(this._pressCache.get(context));
				this._pressCache.delete(context);
				this.emit(`${this.UUID}.singlePress`, evtData);
			}
		});
		// --
	}

	onSinglePress = (callback: (evtData: KeyUpData<unknown>) => void) => this.on(`${this.UUID}.singlePress`, callback);
	onLongPress = (callback: (evtData: KeyDownData<unknown>) => void) => this.on(`${this.UUID}.longPress`, callback);
}

src/propertyInspector/external.js
let globalSettings = {};

window.onload = () => {
	const formEl = document.querySelector('form');
	const customMinEl = document.querySelector('input[name="customMin"]');
	const customMaxEl = document.querySelector('input[name="customMax"]');

	// Load global settings to form
	globalSettings = window.opener.getGlobalSettings();
	customMinEl.value = globalSettings.customMin;
	customMaxEl.value = globalSettings.customMax;

	// Send settigs to property inspector on form edit
	formEl.addEventListener(
		'input',
		Utils.debounce(500, () => {
			window.opener.sendCustomSettingsToInspector({
				customMin: Number(customMinEl.value),
				customMax: Number(customMaxEl.value),
			});
		}),
	);

	// Generate top scores table
	drawTopScores();
};

// Reset top scores listener
document.querySelector('#reset').addEventListener('click', () => {
	const blankTopScores = levels.map(() => Array(times.length).fill(0));
	window.opener.sendCustomSettingsToInspector({ topScores: blankTopScores });
	globalSettings.topScores = blankTopScores;
	drawTopScores();
});

// Draw top scores table
function drawTopScores() {
	const topScores = globalSettings.topScores ?? levels.map(() => Array(times.length).fill(0));
	document.querySelector('#top').innerHTML = createTable(topScores, times, levels);
}

// -- Aux variables/functions to generate top scores table
const times = [5, 10, 20, 30, 'endless'].map((t) => (isNaN(t) ? t : t + 's'));
const levels = ['easy', 'normal', 'hard', 'custom'];

function createTable(data, rowHeadings, columnHeadings) {
	let [...rows] = data;
	if (columnHeadings) {
		rowHeadings = ['', ...rowHeadings];
		rows = rows.map((row, i) => [columnHeadings[i], ...row]);
	}

	return `
	  <table>
		<thead>${getCells(rowHeadings, 'th')}</thead>
		<tbody>${createBody(rows)}</tbody>
	  </table>
	`;
}

function getCells(data, type) {
	return data.map((cell) => `<${type}>${cell}</${type}>`).join('');
}

function createBody(data) {
	return data.map((row) => `<tr>${getCells(row, 'td')}</tr>`).join('');
}
// --


src/plugin/app.ts
import { DidReceiveGlobalSettingsData, KeyDownData, WillAppearData } from './types';
import { SDAction } from './SDAction';

// -- Devices
const devicesList: { [deviceId: string]: number } = {};

// -- Actions and contexts caches
const startAction = new SDAction('dev.theca11.whack-a-mole.start');
const scoreAction = new SDAction('dev.theca11.whack-a-mole.score');
const timerAction = new SDAction('dev.theca11.whack-a-mole.timer');
const tileAction = new SDAction('dev.theca11.whack-a-mole.tile');
const goAction = new SDAction('dev.theca11.whack-a-mole.go-profile');
const startContexts = new Set<string>();
const scoreContexts = new Set<string>();
const timerContexts = new Set<string>();
const tilesContexts = new Set<string>();

// -- Game options config
const times: (number | 'endless')[] = [5, 10, 20, 30, 'endless'];
const levels = ['easy', 'normal', 'hard', 'custom'];
const levelsConfig: Record<string, { min: number, max: number }> = {
	easy: { min: 1000, max: 1500 },
	normal: { min: 650, max: 950 },
	hard: { min: 450, max: 800 },
	custom: { min: 650, max: 950 },
};

// --  Default settings, later updated with saved global settings
type Settings = { level: number, time: number, customMin: number, customMax: number, topScores: number[][] }
let settings: Settings = {
	level: 1,
	time: 1,
	customMin: 500,
	customMax: 1000,
	topScores: levels.map(() => Array(times.length).fill(0)),
};

// -- Main game variables
let countdownInterval: NodeJS.Timeout | undefined; // main game interval
let ready = true;		// ready to start
let inProgress = false;	// game in progress
let score = 0;			// current score
let remainingTime = 0;	// current remaining time
let lastTile: string;	// last tile context with a mole
let showTop = false;	// whether to show last or top score


// -- SD connection and global settings
$SD.onConnected(({ appInfo }: any) => {
	console.log('Stream Deck connected');
	for (const device of appInfo.devices) {
		devicesList[device.id] = device.type;
	}
	$SD.getGlobalSettings();
});

$SD.onDidReceiveGlobalSettings(({ payload }: DidReceiveGlobalSettingsData<Partial<Settings>>) => {
	settings = { ...settings, ...payload.settings };
	levelsConfig.custom = { ...levelsConfig.custom, min: settings.customMin, max: settings.customMax };
	timerContexts.forEach(ctx => setTimerTitle(ctx));
	scoreContexts.forEach(ctx => setScoreTitle(ctx));
});
// --

// -- Actions listeners
// --- Start
startAction.onWillAppear((evtData: WillAppearData<unknown>) => {
	startContexts.add(evtData.context);
	setStartTitle(evtData.context);
	$SD.setState(evtData.context, !inProgress ? 0 : 1);
});

startAction.onWillDisappear((evtData: WillAppearData<unknown>) => {
	startContexts.delete(evtData.context);
});

startAction.onSinglePress(() => {
	if (ready && !inProgress && tilesContexts.size > 0) startGame();
});

startAction.onLongPress((evtData: KeyDownData<unknown>) => {
	if (inProgress) {
		remainingTime = 0;
		endGame();
	}
	else if (ready) {
		$SD.switchToProfile(evtData.device);
	}
});
// ---

// --- Score
scoreAction.onWillAppear((evtData: WillAppearData<unknown>) => {
	scoreContexts.add(evtData.context);
	setScoreTitle(evtData.context);
	$SD.setState(evtData.context, !inProgress ? 0 : 1);
});

scoreAction.onWillDisappear((evtData: WillAppearData<unknown>) => {
	scoreContexts.delete(evtData.context);
});

scoreAction.onSinglePress(() => {
	if (!inProgress) {
		showTop = !showTop;
		scoreContexts.forEach(ctx => setScoreTitle(ctx));
	}
});
// ---

// --- Timer
timerAction.onWillAppear((evtData: WillAppearData<unknown>) => {
	timerContexts.add(evtData.context);
	setTimerTitle(evtData.context);
	$SD.setState(evtData.context, !inProgress ? 0 : 1);
});

timerAction.onWillDisappear((evtData: WillAppearData<unknown>) => {
	timerContexts.delete(evtData.context);
});

timerAction.onSinglePress(() => {
	if (ready && !inProgress) {
		const nextIdx = (settings.time + 1) % times.length;
		settings.time = nextIdx;
		$SD.setGlobalSettings(settings);
		timerContexts.forEach(ctx => setTimerTitle(ctx));
		if (showTop) {
			scoreContexts.forEach(ctx => setScoreTitle(ctx));
		}
	}
});

timerAction.onLongPress(() => {
	if (ready && !inProgress) {
		const nextIdx = (settings.level + 1) % levels.length;
		settings.level = nextIdx;
		$SD.setGlobalSettings(settings);
		timerContexts.forEach(ctx => setTimerTitle(ctx));
		if (showTop) {
			scoreContexts.forEach(ctx => setScoreTitle(ctx));
		}
	}
});
// ---

// --- Tile
tileAction.onWillAppear((evtData: WillAppearData<unknown>) => {
	tilesContexts.add(evtData.context);
	$SD.setState(evtData.context, 0);
});

tileAction.onWillDisappear((evtData: WillAppearData<unknown>) => {
	tilesContexts.delete(evtData.context);
	if (inProgress && tilesContexts.size === 0) endGame();
});

tileAction.onSinglePress((evtData: KeyDownData<unknown>) => {
	const { state } = evtData.payload;
	if (state) {
		$SD.setState(evtData.context, 0);
		score += getMoleMultiplier(state);
		scoreContexts.forEach(ctx => setScoreTitle(ctx));
		if (times[settings.time] === 'endless') {
			remainingTime += getMoleMultiplier(state);
			timerContexts.forEach(ctx => setTimerTitle(ctx));
		}
	}
});
// ---

// --- Go to profile
function getProfilePath(deviceType: number) {
	switch (deviceType) {
		case 0:
		case 3:
			return 'profiles/SD';
		case 1:
			return 'profiles/SDMini';
		case 2:
			return 'profiles/SDXL';
		case 7:
			return 'profiles/SDPlus';
		default:
			return '';
	}
}

goAction.onKeyDown((evtData: KeyDownData<unknown>) => {
	const { device, context } = evtData;
	const profile = getProfilePath(devicesList[device]);
	if (profile) {
		$SD.switchToProfile(device, profile);
	}
	else {	// set temporal unsupported message
		$SD.setTitle(context, 'DEVICE\nNOT\nSUPPORTED');
		setTimeout(() => $SD.setTitle(context, ''), 2000);
	}
});
// ---
// --

// -- Whack-a-mole methods
function getMoleKind() {
	const rnd = Math.floor(Math.random() * 10);
	return (rnd === 2 || rnd === 3) ? rnd : 1;
}

function getMoleMultiplier(moleKind: number) {
	return (moleKind === 1 || moleKind === 2) ? moleKind : -1;
}

function getRandomNumber(min: number, max: number) {
	return Math.round(Math.random() * (max - min) + min);
}

function getRandomTile() {
	const idx = Math.floor(Math.random() * tilesContexts.size);
	const tile = [...tilesContexts][idx];
	if (tile === lastTile && tilesContexts.size > 1) {
		return getRandomTile();
	}
	lastTile = tile;
	return tile;
}

function peep() {
	const lengthSettings = levelsConfig[levels[settings.level]];
	const peepLength = getRandomNumber(lengthSettings.min, lengthSettings.max);
	const tile = getRandomTile();
	$SD.setState(tile, getMoleKind());
	setTimeout(() => {
		$SD.setState(tile, 0);
		if (inProgress) peep();
	}, peepLength);
}

function startGame() {
	ready = false;
	inProgress = true;
	startContexts.forEach(ctx => $SD.setState(ctx, 1));
	timerContexts.forEach(ctx => $SD.setState(ctx, 1));
	scoreContexts.forEach(ctx => $SD.setState(ctx, 1));
	startContexts.forEach(ctx => setStartTitle(ctx));
	score = 0;
	showTop = false;
	scoreContexts.forEach(ctx => setScoreTitle(ctx));
	remainingTime = times[settings.time] === 'endless' ? 10 : times[settings.time] as number;
	timerContexts.forEach(ctx => setTimerTitle(ctx));
	countdownInterval = setInterval(() => {
		remainingTime -= 1;
		if (remainingTime === 0) {
			endGame();
		}
		timerContexts.forEach(ctx => setTimerTitle(ctx));
	}, 1000);
	peep();
}

function endGame() {
	inProgress = false;
	clearInterval(countdownInterval);
	$SD.setState(lastTile, 0);
	startContexts.forEach(ctx => $SD.setState(ctx, 0));
	timerContexts.forEach(ctx => $SD.setState(ctx, 0));
	scoreContexts.forEach(ctx => $SD.setState(ctx, 0));
	startContexts.forEach(ctx => setStartTitle(ctx));
	const currentTopScore = settings.topScores[settings.level][settings.time];
	if (score > currentTopScore) {
		settings.topScores[settings.level][settings.time] = score;
		$SD.setGlobalSettings(settings);
	}
	setTimeout(() => {
		ready = true;
		startContexts.forEach(ctx => setStartTitle(ctx));
		timerContexts.forEach(ctx => setTimerTitle(ctx));
	}, 2500);
}

function setStartTitle(context: string) {
	if (!inProgress) {
		if (ready) {
			$SD.setTitle(context, 'START\n-\nHold\nto EXIT');
		}
		else {
			$SD.setTitle(context, 'GAME\nOVER');
		}
	}
	else {
		$SD.setTitle(context, 'WHACK!');
	}
}

function setScoreTitle(context: string) {
	if (!showTop) {
		$SD.setTitle(context, `SCORE\n\n${score}`);
	}
	else {
		$SD.setTitle(context, `TOP\nSCORE\n\n${settings.topScores[settings.level][settings.time]}`);
	}
}

function setTimerTitle(context: string) {
	const time = (inProgress || !ready) ? remainingTime : times[settings.time];
	if (time === 'endless') {
		$SD.setTitle(context, `${levels[settings.level].toLocaleUpperCase()}\n\nEndless`);
	}
	else {
		$SD.setTitle(context, `${levels[settings.level].toLocaleUpperCase()}\n\n${formatTime(time)}`);
	}
}

function formatTime(seconds: number) {
	const minutes = Math.floor(seconds / 60).toLocaleString(undefined, { minimumIntegerDigits: 2 });
	const secs = (seconds % 60).toLocaleString(undefined, { minimumIntegerDigits: 2 });
	return `${minutes}:${secs}`;
}
// --